### PR TITLE
[New features] Add three MoE features: HashRouter (topk_method=hash), SftPlus score, ClampedSwiGLU

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
### PR Category
Operator Mechanism

### PR Types
New features

### Description
Add three MoE features from Megatron-LM:

**1. ClampedSwiGLU** (`activation_func_clamp_value`)
Adds `clamp(x, max=clamp_value)` after SwiGLU activation to stabilize expert gradient magnitudes. The existing `WeightedSwiGLUFunction` is extended with an optional `clamp_value` parameter; when `None` (default), behavior is identical to before.

**2. SftPlus score normalization** (`scoring_func="sftplus"`)
Replaces softmax with `softplus(x) / sum(softplus(x))` for router score computation, producing non-negative gate scores without the winner-take-all pressure of softmax.

**3. Hash routing** (`moe_router_topk_method: hash`)
Deterministic expert assignment via `(token_id + k) % num_experts`, bypassing all learned gate computation and aux-loss. Useful for ablation studies and debugging load balance.

**Architecture**: `HashRouter` has been merged into `TopKRouter` as a `topk_method="hash"` branch, reducing ~160 lines of duplicated code. The hash branch short-circuits gate matmul and returns the same 8-tuple output format.

**Config**:
```yaml
moe_router_topk_method: hash          # or via --moe-n-hash-layers N (last N layers)
moe_hash_router_pad_token_id: 0       # set to tokenizer pad_token_id if != 0
```

**Tests**: 48 router unit tests passing, pre-commit checks clean.

### CODE REVIEW

**Comment by adam-xiaoyao** [comment]
Thanks for the review. Addressed in commit 8d34f50:

- 🔴 **`moe_router.py:951` sequence_parallel token-id ordering** — fixed by transposing `input_ids` from `[B,S]` to `[S,B]` before flatten when `sequence_parallel=True`. Added regression test `test_sequence_parallel_token_alignment`.
- ❓ **`moe_router.py:940` assert vs ValueError** — replaced `assert input_ids is not None` with explicit `raise ValueError`; updated test to expect `ValueError`.
- 🟡 **`fused_bias_swiglu.py:337` dead code** — removed `ClampedWeightedSwiGLUFunction`; `WeightedSwiGLUFunction` already supports the clamp path via its optional `clamp_value` argument.
- 📝 **PR title / body** — title prefixed with `[Feature]`; body restructured to the `### PR Category` / `### PR Types` / `### Description` template.
- ➕ **Bonus fix** — `WeightedSwiGLUFunction.backward` previously returned 4 values (`tmp, wgrad, None, None`), which violates Paddle's PyLayer arity rule (one gradient per tensor input). Reduced to `return tmp, wgrad`. Added an end-to-end `paddle.grad` test for the clamp path.

Note on `arguments.py` / `yaml_arguments.py`: those files don't enumerate per-field declarations — `arguments.py` iterates over `dataclasses.fields(TransformerConfig)` and `yaml_arguments.py` flattens any DictConfig — so `moe_n_hash_layers` is auto-picked up via the new `TransformerConfig` field. No change needed unless an explicit CLI alias is desired.

All 76 tests pass; coverage on PR-added lines is **92.9%**.

**Comment by adam-xiaoyao** [comment]
Thanks for the detailed review. Addressed in the latest commits:

---

**Fixed in commit `cef227c`:**

- **`masked_fill` scalar vs Tensor** — changed `paddle.to_tensor(-1, dtype=paddle.int64)` to the literal `-1`. This matches the existing `masked_fill(-1)` call at line 817 and avoids any Paddle version sensitivity with Tensor-valued `fill_value`.

- **`moe_n_hash_layers` CLI registration** — added `--moe-n-hash-layers` to `training/arguments.py`. Note that YAML was already working (all YAML keys flow through `_flatten_configs` → `core_transformer_config_from_args`). The CLI gap is now closed too.

- **`pad_token_id == 0` assumption** — added a `.. note::` block in the `HashRouter` docstring explicitly documenting this constraint and flagging `moe_hash_router_pad_token_id` as a future config field. Matching Megatron-LM behaviour for now is deliberate; a follow-up can generalise if needed.

- **PR title** — updated to `[New features]` to match the repo convention (e.g. #915).

---

**No change needed / already addressed (earlier commits):**

- **`HashRouter` can't be activated in standard path** — the activation path is fully wired: `MoELayer.__init__` selects `HashRouter` when `moe_n_hash_layers > 0` and `layer_number >= num_hidden_layers - moe_n_hash_layers`; `MoELayer.forward` → `compute_gate` → `self.gate(hidden_states, input_ids=input_ids)` passes `input_ids` through. See `moe_layer.py:692` and `moe_layer.py:848`.

- **`sequence_parallel` token-order bug** — fixed in `8d34f50` with `input_ids.transpose([1,0]).reshape([-1])` when `sequence_parallel=True`.

- **`assert` → `ValueError`** — fixed in `8d34f50`.

- **`ClampedWeightedSwiGLUFunction` dead code** — deleted in `8d34f50`.

- **PyLayer backward arity** — fixed to `return tmp, wgrad` in `8d34f50`.

**Comment by adam-xiaoyao** [comment]
Thanks for the careful checklist review. **YAML is already fully supported** — no changes to `yaml_arguments.py` are needed.

Here is how the YAML path works end-to-end:

1. `parse_args` detects `--configs yaml_file.yaml` and calls `load_yaml(path)`.
2. `yaml_arguments.load_yaml` calls `_flatten_configs`, which **recursively flattens all keys** from the nested DictConfig into a single flat OmegaConf object — regardless of field name.
3. `core_transformer_config_from_args` then does `hasattr(args, f.name)` for every `dataclasses.fields(TransformerConfig)` field, picking up whatever the YAML contains.

In other words, `yaml_arguments.py` is **field-name-agnostic**: as long as the user writes the correct key in their YAML, it is automatically forwarded to `TransformerConfig` without any change to `yaml_arguments.py`.

Verified locally:

```yaml
# user's config.yaml
model:
  moe_n_hash_layers: 4
  moe_hash_router_pad_token_id: 2
```

```python
from paddlefleet.training.yaml_arguments import load_yaml
args = load_yaml("config.yaml")
# hasattr(args, 'moe_n_hash_layers')        → True, value = 4
# hasattr(args, 'moe_hash_router_pad_token_id') → True, value = 2
```

The `yaml_arguments.py` change would only be necessary if it had a hardcoded allow-list of field names (like some frameworks do). PaddleFleet's design deliberately avoids that, so every `TransformerConfig` field is automatically YAML-accessible.

I'll add a note to the PR description to make this explicit for future readers.

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 1]
```diff
@@ -873,3 +875,147 @@ def forward(self, input, input_ids=None):
...
+                - l_aux (None)
+                - l_zloss (None)
+        """
+        assert input_ids is not None, (
+            "HashRouter requires input_ids to be provided. "
+            "Make sure input_ids is passed through the model forward to the MoE layer."
+        )
+
+        if len(input.shape) != 3:
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: Fixed. When `sequence_parallel=True`, `input_ids` is now transposed before flattening: `input_ids.transpose([1, 0]).reshape([-1])` to match the `[S, B, H]` token order. See `TopKRouter.forward` hash branch.

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 2]
```diff
@@ -873,3 +875,147 @@ def forward(self, input, input_ids=None):
...
+                when sequence_parallel=True). Not used for routing itself.
+            input_ids (paddle.Tensor): Token IDs, shape [B, S]. **Required.**
+
+        Returns:
+            tuple: 8-element tuple matching TopKRouter output format:
+                - capacity (None)
+                - top_gate  [B*S, K]  per-token expert weights
+                - top_idx   [B*S, K]  selected expert indices (-1 for padding)
+                - probs     [B*S, E]  sparse weight matrix
+                - mask      [B*S, E]  binary routing mask
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: Fixed. Replaced `assert` with explicit `raise ValueError` in the hash routing branch.

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 3]
```diff
@@ -873,3 +875,147 @@ def forward(self, input, input_ids=None):
...
+                - token_priority (None)
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: Fixed. Replaced `assert` with explicit `raise ValueError` in the hash routing branch.

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 4]
```diff
@@ -873,3 +875,156 @@ def forward(self, input, input_ids=None):
...
+                - l_aux (None)
+                - l_zloss (None)
+        """
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: Fixed. Replaced `assert` with explicit `raise ValueError` in the hash routing branch.

File: src/paddlefleet/fusions/fused_bias_swiglu.py [hunk 5]
```diff
@@ -214,20 +306,47 @@ def backward(ctx, grad_output):
...
+        if clamp_value is not None:
+            tmp, wgrad = clamped_weighted_swiglu_back(
+                grad_output, input, weights, clamp_value
+            )
+        else:
+            tmp, wgrad = weighted_swiglu_back(grad_output, input, weights)
+        return tmp, wgrad, None, None
+
+
+class ClampedWeightedSwiGLUFunction(paddle.autograd.PyLayer):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/fusions/fused_bias_swiglu.py]: Fixed. `ClampedWeightedSwiGLUFunction` has been deleted. The clamped path is handled entirely through `WeightedSwiGLUFunction` via the existing `clamp_value` parameter.

File: src/paddlefleet/fusions/fused_bias_swiglu.py [hunk 6]
```diff
@@ -214,20 +306,47 @@ def backward(ctx, grad_output):
...
+        ctx.clamp_value = clamp_value
+        return clamped_weighted_swiglu(input, weights, clamp_value)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input, weights = ctx.saved_tensor()
+        tmp, wgrad = clamped_weighted_swiglu_back(
+            grad_output, input, weights, ctx.clamp_value
+        )
         return tmp, wgrad, None
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/fusions/fused_bias_swiglu.py]: Fixed. `ClampedWeightedSwiGLUFunction` has been deleted. The clamped path is handled entirely through `WeightedSwiGLUFunction` via the existing `clamp_value` parameter.

File: src/paddlefleet/fusions/fused_bias_swiglu.py [hunk 7]
```diff
@@ -214,20 +308,47 @@ def backward(ctx, grad_output):
...
+        if clamp_value is not None:
+            tmp, wgrad = clamped_weighted_swiglu_back(
+                grad_output, input, weights, clamp_value
+            )
+        else:
+            tmp, wgrad = weighted_swiglu_back(grad_output, input, weights)
+        return tmp, wgrad, None, None
+
+
+class ClampedWeightedSwiGLUFunction(paddle.autograd.PyLayer):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/fusions/fused_bias_swiglu.py]: Fixed. `ClampedWeightedSwiGLUFunction` has been deleted. The clamped path is handled entirely through `WeightedSwiGLUFunction` via the existing `clamp_value` parameter.

File: src/paddlefleet/transformer/transformer_config.py [hunk 8]
```diff
@@ -462,6 +463,11 @@ class TransformerConfig(ModelParallelConfig):
     moe_router_force_load_balancing: bool = False
     """Force load balancing with random logits for MoE router."""
 
+    moe_n_hash_layers: int = 0
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:466]: Fixed for CLI via `arguments.py` (added `--moe-n-hash-layers` and `--moe-hash-router-pad-token-id`). For YAML: `yaml_arguments.py` is a pass-through loader (`OmegaConf.load` + flatten) that forwards all keys directly to the config—no explicit field registration needed. Users can simply add `moe_n_hash_layers: 4` / `moe_hash_router_pad_token_id: 0` to their YAML config.

[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:466]: Fixed for CLI via `arguments.py` (added `--moe-n-hash-layers` and `--moe-hash-router-pad-token-id`). For YAML: `yaml_arguments.py` is a pass-through loader (`OmegaConf.load` + flatten) that forwards all keys directly to the config—no explicit field registration needed. Users can simply add `moe_n_hash_layers: 4` / `moe_hash_router_pad_token_id: 0` to their YAML config.

[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:466]: Fixed for CLI via `arguments.py` (added `--moe-n-hash-layers` and `--moe-hash-router-pad-token-id`). For YAML: `yaml_arguments.py` is a pass-through loader (`OmegaConf.load` + flatten) that forwards all keys directly to the config—no explicit field registration needed. Users can simply add `moe_n_hash_layers: 4` / `moe_hash_router_pad_token_id: 0` to their YAML config.

[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:466]: Fixed for CLI via `arguments.py` (added `--moe-n-hash-layers` and `--moe-hash-router-pad-token-id`). For YAML: `yaml_arguments.py` is a pass-through loader (`OmegaConf.load` + flatten) that forwards all keys directly to the config—no explicit field registration needed. Users can simply add `moe_n_hash_layers: 4` / `moe_hash_router_pad_token_id: 0` to their YAML config.

File: src/paddlefleet/transformer/transformer_config.py [hunk 9]
```diff
@@ -462,6 +463,17 @@ class TransformerConfig(ModelParallelConfig):
...
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:466]: Fixed for CLI via `arguments.py` (added `--moe-n-hash-layers` and `--moe-hash-router-pad-token-id`). For YAML: `yaml_arguments.py` is a pass-through loader (`OmegaConf.load` + flatten) that forwards all keys directly to the config—no explicit field registration needed. Users can simply add `moe_n_hash_layers: 4` / `moe_hash_router_pad_token_id: 0` to their YAML config.

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 10]
```diff
@@ -138,6 +138,7 @@ def __init__(
         config: TransformerConfig,
         sublayers: MoESublayers | None = None,
         pg_collection: ProcessGroupCollection | None = None,
+        layer_number: int | None = None,
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py:141]: The `input_ids` forwarding chain is already in place: `MoELayer.forward` accepts `input_ids` and passes it to `self.gate(input, input_ids=input_ids)`. The `TopKRouter.forward` hash branch then validates and uses it.

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 11]
```diff
@@ -873,3 +875,165 @@ def forward(self, input, input_ids=None):
...
+        top_gate = paddle.full(
+            [num_tokens, self.num_experts_per_tok],
+            weight_val,
+            dtype=paddle.float32,
+        )  # [B*S, K]
+
+        # ------------------------------------------------------------------ #
+        # Padding mask: token_id == 0 → weight=0, idx=-1                      #
+        # ------------------------------------------------------------------ #
+        valid_mask = (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: Fixed. Changed to scalar `-1` in commit cef227c.

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 12]
```diff
@@ -873,3 +875,165 @@ def forward(self, input, input_ids=None):
...
+            flat_ids = input_ids.reshape([-1]).cast(paddle.int64)
+
+        num_tokens = batch_size * seq_len
+
+        # ------------------------------------------------------------------ #
+        # Hash assignment: expert_k = (token_id + k) % num_experts            #
+        # ------------------------------------------------------------------ #
+
+        # offsets: [[0, 1, ..., K-1]]  ->  broadcast with flat_ids
+        offsets = paddle.arange(
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: Fixed. Added `moe_hash_router_pad_token_id` field to `TransformerConfig` (default 0). Users can set it to their tokenizer's actual pad token id, e.g. LLaMA's `eos_token_id`.

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 13]
```diff
@@ -224,7 +225,29 @@ def __init__(
...
+            )
+            logger.info(
+                f"MoELayer layer_number={layer_number}: using HashRouter "
+                f"(moe_n_hash_layers={config.moe_n_hash_layers}, "
+                f"num_hidden_layers={config.num_hidden_layers})"
+            )
+        else:
+            self.gate = TopKRouter(config=config, pg_collection=pg_collection)
+            if layer_number is not None:
+                self.gate.set_layer_number(layer_number)
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py]: Confirmed: `TopKRouter.set_layer_number` exists (inherited from `StandardMoERouter` and overridden in `TopKRouter`). After the HashRouter → TopKRouter refactoring, `MoELayer.set_layer_number` now sets `self.gate.topk_method = "hash"` directly on the existing `TopKRouter` instance instead of replacing it.

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 14]
```diff
@@ -138,6 +138,7 @@ def __init__(
         config: TransformerConfig,
         sublayers: MoESublayers | None = None,
         pg_collection: ProcessGroupCollection | None = None,
+        layer_number: int | None = None,
     ):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py:142]: By design: when `layer_number` is not provided, `MoELayer` defaults to TopKRouter behavior (no hash routing), which is the correct fallback for any caller that doesn't set layer numbers. In the standard training pipeline, `set_layer_number` is called during model construction when layer specs are assembled. If `moe_n_hash_layers == 0` (default), hash routing is never activated regardless.
